numpy/core/src/npysort/selection.cpp

```diff
@@ -377,269 +378,105 @@ introselect_(type *v, npy_intp *tosort, npy_intp num, npy_intp kth,
-                                                 npiv);
-}
-NPY_NO_EXPORT int
-aintroselect_ulonglong(npy_ulonglong *v, npy_intp *tosort, npy_intp num,
-                       npy_intp kth, npy_intp *pivots, npy_intp *npiv,
-                       void *NOT_USED)
-{
-    return introselect_<npy::ulonglong_tag, true>(v, tosort, num, kth, pivots,
-                                                  npiv);
-}
-NPY_NO_EXPORT int
-aintroselect_half(npy_half *v, npy_intp *tosort, npy_intp num, npy_intp kth,
-                  npy_intp *pivots, npy_intp *npiv, void *NOT_USED)
-{
-    return introselect_<npy::half_tag, true>(v, tosort, num, kth, pivots,
-                                             npiv);
-}
-NPY_NO_EXPORT int
-aintroselect_float(npy_float *v, npy_intp *tosort, npy_intp num, npy_intp kth,
-                   npy_intp *pivots, npy_intp *npiv, void *NOT_USED)
+template <typename Tag>
+static int
+introselect_arg(void *v, npy_intp *tosort, npy_intp num, npy_intp kth,
+                npy_intp *pivots, npy_intp *npiv, void *)
 {
-    return introselect_<npy::float_tag, true>(v, tosort, num, kth, pivots,
-                                              npiv);
+    return introselect_<Tag, true>((typename Tag::type *)v, tosort, num, kth,
+                                   pivots, npiv);
 }
-NPY_NO_EXPORT int
-aintroselect_double(npy_double *v, npy_intp *tosort, npy_intp num,
-                    npy_intp kth, npy_intp *pivots, npy_intp *npiv,
-                    void *NOT_USED)
+
+struct arg_map {
+    int typenum;
+    PyArray_PartitionFunc *part[NPY_NSELECTS];
+    PyArray_ArgPartitionFunc *argpart[NPY_NSELECTS];
+};
+
+template <class... Tags>
+static constexpr std::array<arg_map, sizeof...(Tags)>
+make_partition_map(npy::taglist<Tags...>)
 {
-    return introselect_<npy::double_tag, true>(v, tosort, num, kth, pivots,
-                                               npiv);
+    return std::array<arg_map, sizeof...(Tags)>{
+            arg_map{Tags::type_value, &introselect_noarg<Tags>,
+                    &introselect_arg<Tags>}...};
 }
-NPY_NO_EXPORT int
-aintroselect_longdouble(npy_longdouble *v, npy_intp *tosort, npy_intp num,
-                        npy_intp kth, npy_intp *pivots, npy_intp *npiv,
-                        void *NOT_USED)
+
+struct partition_t {
+    using taglist =
+            npy::taglist<npy::bool_tag, npy::byte_tag, npy::ubyte_tag,
+                         npy::short_tag, npy::ushort_tag, npy::int_tag,
+                         npy::uint_tag, npy::long_tag, npy::ulong_tag,
+                         npy::longlong_tag, npy::ulonglong_tag, npy::half_tag,
+                         npy::float_tag, npy::double_tag, npy::longdouble_tag,
+                         npy::cfloat_tag, npy::cdouble_tag,
+                         npy::clongdouble_tag>;
+
+    static constexpr std::array<arg_map, taglist::size> map =
+            make_partition_map(taglist());
```

Comment:
@charris : this is the key part: Initialization of the array based on a tag list. `make_partition_map` is in charge of generating the array, using the same template function for each element of the taglist. A taglist is just a list of types